### PR TITLE
Fix bug in multi-gpu

### PR DESCRIPTION
**Pull request by seed93**
Rewrite DataParallelTable to fit RNN model for multi-gpu.

TO DO:
1. Acceleration
2. Check why the result of multi-gpu is different from the result of single gpu.

Please don't merge until we finish this.

### CODE REVIEW

**Comment by SeanNaren** [comment]
Hey let me know if you need any help, thanks :)

**Comment by SeanNaren** [comment]
Hey, if I recall correctly the problem of multi-gpu is around the fact that the output of the network is 2D not 3D (seqLength x batchSize x feat). Is this correct? If so I might have a solution that may be more elegant.

**Comment by seed93** [comment]
Yes, its output is 2D. Currently we think variable lengths in minibatch lead to many problems and we want to use same lengths. We are still working, good luck.

**Comment by SeanNaren** [comment]
@seed93 Have a look at [this](https://gist.github.com/SeanNaren/a271d1f4af5fe57594b0ece363b01dec) module, it will handle converting `T*N*H` to `(T*N)*H` and revert it back (so the output will be `T*N*H`).

This can be used on the final batch norm and fully connected so that the output of the network is still a 3D tensor, and it works well from what I can see. The architecture has changed a bit from the main branch otherwise I could implement it for you!

**Comment by seed93** [comment]
Thank you! We are trying to remove mask, hoping to get a good result. Any progress will be reported. I suggest you not to merge this branch now. Sorry for that

**Comment by SeanNaren** [comment]
@seed93 no worries, I've actually already started removing the mask, I was going to open an issue to explain why. Baidu doesn't handle masking the way we have and just makes sure that they are ignored in the loss calculations. I've been experimenting to see how this works out, and it seems to definitely take a lot less time to train. This is shown in [this](https://github.com/SeanNaren/CTCSpeechRecognition/tree/NoMasking) branch :)

**Comment by jianminsun** [comment]
@seed93 Got errors when I tried to save the model in the case of nGPU>1. Is there a way to fix this? 
Saving model..
/home/ubuntu/torch/install/bin/luajit: ...u/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:630: attempt to index field 'modules' (a nil value)
stack traceback:
    ...u/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:630: in function 'applyChanges'
    ...u/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:644: in function 'setup'
    ...u/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:651: in function 'exec'
    ...u/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:159: in function 'clearState'
    ./UtilsMultiGPU.lua:55: in function 'saveDataParallel'
    ./Network.lua:245: in function 'saveNetwork'
    ./Network.lua:234: in function 'trainNetwork'
    AN4CTCTrain.lua:46: in main chunk
    [C]: in function 'dofile'
    ...untu/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
    [C]: at 0x00406670
